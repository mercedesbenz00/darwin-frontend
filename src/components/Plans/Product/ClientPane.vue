<template>
  <div class="client-pane">
    <header-3>Your Annotation plan is managed by {{ partnerName }}</header-3>
    <paragraph-14>
      You can take over billing for this plan anytime.
      This will allow you to view your usage, storage, and select a different plan.
      Some automation features may be locked by {{ partnerName }}.
    </paragraph-14>
  </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'
import { State } from 'vuex-class'

import Header3 from '@/components/Common/Header3.vue'
import Paragraph14 from '@/components/Common/Paragraph14.vue'
import { RootState, TeamPayload } from '@/store/types'

@Component({
  name: 'client-pane',
  components: {
    Header3,
    Paragraph14
  }
})
export default class ClientPane extends Vue {
  @State((state: RootState) => state.team.currentTeam)
  currentTeam!: TeamPayload

  get partnerName (): string {
    const { currentTeam } = this
    if (!currentTeam.partner) { throw new Error('Trying to render client UI for non-client team') }
    return currentTeam.partner.name
  }
}
</script>

<style lang="scss" scoped>
.client-pane {
  @include col--center;
  row-gap: 20px;
  min-height: 300px;
  text-align: center;

  > * {
    max-width: 620px;
  }
}
</style>
