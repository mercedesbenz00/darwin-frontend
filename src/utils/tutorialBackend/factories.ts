import {
  AnnotateStagePayload,
  CodeStagePayload,
  CompleteStagePayload,
  DatasetImagePayload,
  DatasetItemPayload,
  DatasetItemStatus,
  DatasetItemType,
  ImagePayload,
  ModelStagePayload,
  ReviewStagePayload,
  StageType,
  WorkflowStagePayload,
  WorkflowStageTemplatePayload,
  WorkflowTemplatePayload
} from '@/store/types'

import TutorialError from './TutorialError'
import { dataset } from './data/dataset'

const itemId = (datasetImageId: number) => datasetImageId * 10
const workflowId = (datasetImageId: number) => datasetImageId * 100

const stageId = (datasetImageId: number, stageNumber: number) =>
  workflowId(datasetImageId) + stageNumber

const initializeDatasetImage = (image: ImagePayload): DatasetImagePayload => ({
  dataset_id: dataset.id,
  dataset_video_id: null,
  id: image.id * 10,
  image,
  seq: image.id,
  set: 1
})

const initializeItem = (datasetImage: DatasetImagePayload): DatasetItemPayload => ({
  archived_reason: null,
  archived: false,
  current_workflow: null,
  current_workflow_id: null,
  dataset_id: datasetImage.dataset_id,
  dataset_image_id: datasetImage.id,
  dataset_image: datasetImage,
  dataset_video_id: null,
  dataset_video: null,
  file_size: 0,
  filename: datasetImage.image.original_filename,
  height: datasetImage.image.height,
  id: itemId(datasetImage.id),
  inserted_at: '2020-06-12T00:00:00',
  labels: [],
  path: '/',
  priority: 0,
  seq: datasetImage.seq,
  set: datasetImage.set,
  status: DatasetItemStatus.annotate,
  updated_at: '2020-06-12T00:00:00',
  width: datasetImage.image.width,
  type: DatasetItemType.image
})

const initializeStage = (
  datasetImage: DatasetImagePayload,
  stageTemplate: WorkflowStageTemplatePayload
): WorkflowStagePayload => {
  const { type } = stageTemplate
  const payload = {
    assignee_id: null,
    completed: false,
    completes_at: null,
    dataset_item_id: itemId(datasetImage.id),
    id: stageId(datasetImage.id, stageTemplate.stage_number),
    metadata: {},
    number: stageTemplate.stage_number,
    skipped_reason: null,
    skipped: false,
    template_metadata: stageTemplate.metadata,
    type: stageTemplate.type,
    workflow_id: workflowId(datasetImage.id),
    workflow_stage_template_id: stageTemplate.id
  }

  if (type === StageType.Annotate) { return payload as AnnotateStagePayload }
  if (type === StageType.Code) { return payload as CodeStagePayload }
  if (type === StageType.Complete) { return payload as CompleteStagePayload }
  if (type === StageType.Model) { return payload as ModelStagePayload }
  if (type === StageType.Review) { return payload as ReviewStagePayload }

  throw new Error('Invalid type given')
}

const initializeWorkflow = (
  datasetImage: DatasetImagePayload,
  workflowTemplate: WorkflowTemplatePayload,
  currentStage: number
): Pick<DatasetItemPayload, 'current_workflow' | 'current_workflow_id'> => {
  const annotateTemplate =
    workflowTemplate.workflow_stage_templates.find(t => t.type === StageType.Annotate)

  const reviewTemplate =
    workflowTemplate.workflow_stage_templates.find(t => t.type === StageType.Review)

  const completeTemplate =
    workflowTemplate.workflow_stage_templates.find(t => t.type === StageType.Complete)

  if (!annotateTemplate || !reviewTemplate || !completeTemplate) {
    throw new TutorialError('Initializing workflow using invalid template ')
  }

  const id = workflowId(datasetImage.id)

  const workflow: DatasetItemPayload['current_workflow'] = {
    dataset_item_id: itemId(datasetImage.id),
    id,
    workflow_template_id: workflowTemplate.id,
    current_stage_number: currentStage,
    current_workflow_stage_template_id: annotateTemplate.id,
    status: DatasetItemStatus.annotate,
    stages: {
      1: [initializeStage(datasetImage, annotateTemplate)],
      2: [initializeStage(datasetImage, reviewTemplate)],
      3: [initializeStage(datasetImage, completeTemplate)]
    }
  }

  return { current_workflow: workflow, current_workflow_id: id }
}

export const createItem = (
  image: ImagePayload,
  workflowTeplate: WorkflowTemplatePayload,
  stageNumber: number = 1
): DatasetItemPayload => {
  const datasetImage = initializeDatasetImage(image)
  return {
    ...initializeItem(datasetImage),
    ...initializeWorkflow(datasetImage, workflowTeplate, stageNumber)
  }
}
