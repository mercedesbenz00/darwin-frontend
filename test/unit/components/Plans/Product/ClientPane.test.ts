import { createLocalVue, shallowMount } from '@vue/test-utils'
import Vuex from 'vuex'

import createTestStore from 'test/unit/createTestStore'
import { buildTeamPayload } from 'test/unit/factories'

import ClientPane from '@/components/Plans/Product/ClientPane.vue'

const localVue = createLocalVue()
localVue.use(Vuex)

let store: ReturnType<typeof createTestStore>

const itMatchesSnapshot = () => it('matches snapshot', () => {
  const wrapper = shallowMount(ClientPane, { localVue, store })
  expect(wrapper).toMatchSnapshot()
})

beforeEach(() => {
  store = createTestStore()
  const partner = buildTeamPayload({ name: 'Some partner' })
  const client = buildTeamPayload({ id: 1, managed_status: 'client', partner })
  store.commit('team/SET_CURRENT_TEAM', client)
})

itMatchesSnapshot()
