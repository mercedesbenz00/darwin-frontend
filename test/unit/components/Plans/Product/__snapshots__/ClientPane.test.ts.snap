// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div class="client-pane">
  <header-3-stub>Your Annotation plan is managed by Some partner</header-3-stub>
  <paragraph-14-stub>
    You can take over billing for this plan anytime.
    This will allow you to view your usage, storage, and select a different plan.
    Some automation features may be locked by Some partner.
  </paragraph-14-stub>
</div>
`;
